File: clientRust/src/gui.rs
//! Machine-readable JSON progress interface for the `open-rmbt-desktop` app.
//!
//! See `doc/json_interface.md` for the full specification. All JSON output is
//! gated by the `-v` flag; when disabled these functions are no-ops (except the
//! progress counter, which is a cheap atomic used by the interim monitor).
//!
//! Throughput is reported in **decimal** Mbit/s (`bytes * 8 / seconds / 1e6`),
//! ping RTTs in **ms**, and `pingTimeNs` in **ns**, per the spec.

use serde_json::json;
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::thread::{self, JoinHandle};
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

static ENABLED:  AtomicBool = AtomicBool::new(false);
/// Cumulative bytes transferred in the current transfer phase (all threads).
static PROGRESS: AtomicU64  = AtomicU64::new(0);

#[derive(Clone, Copy)]
pub enum Direction { Down, Up }

pub fn set_enabled(v: bool) { ENABLED.store(v, Ordering::Relaxed); }
pub fn enabled() -> bool    { ENABLED.load(Ordering::Relaxed) }

fn now_ms() -> u64 {
    SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_millis() as u64
}

fn emit(v: serde_json::Value) {
    if enabled() { println!("{v}"); }
}

// ─── Plain-text lifecycle sentinels (backward-compat shim, gui mode only) ───────

pub fn starting_test() { if enabled() { println!("STARTING TEST."); } }
pub fn ending_test()   { if enabled() { println!("ENDING TEST."); } }

// ─── Messages ───────────────────────────────────────────────────────────────

pub fn state_change(state: &str) {
    emit(json!({ "type": "STATE_CHANGE", "time": now_ms(), "state": state }));
}

pub fn uuid_info(test_uuid: Option<&str>, open_test_uuid: Option<&str>, token: &str) {
    emit(json!({
        "type":         "UUID_INFO",
        "testUuid":     test_uuid,
        "openTestUuid": open_test_uuid,
        "testToken":    token,
    }));
}

pub fn ping_result(client_ns: u64, server_ns: u64, time_ns: u64) {
    emit(json!({
        "type":       "PING_RESULT",
        "time":       now_ms(),
        "phase":      "PING",
        "pingTimeNs": time_ns,
        "pingClient": client_ns as f64 / 1e6, // ns → ms
        "pingServer": server_ns as f64 / 1e6, // ns → ms
        "status":     "PING",
    }));
}

fn transfer_result(dir: Direction, bytes: u64, mbps: f64, open_test_uuid: Option<&str>, progress_frac: f64) {
    let (typ, phase) = match dir {
        Direction::Down => ("DOWNLOAD_RESULT", "DOWN"),
        Direction::Up   => ("UPLOAD_RESULT",   "UP"),
    };
    let (down, up) = match dir {
        Direction::Down => (mbps, -1.0),
        Direction::Up   => (-1.0, mbps),
    };
    emit(json!({
        "type":         typ,
        "time":         now_ms(),
        "phase":        phase,
        "down":         down,
        "up":           up,
        "bytes":        bytes,
        "testOpenUuid": open_test_uuid,
        "progress":     progress_frac,
        "status":       phase,
    }));
}

// ─── Interim throughput monitor ───────────────────────────────────────────────

pub fn add_progress(bytes: u64) { PROGRESS.fetch_add(bytes, Ordering::Relaxed); }

/// Handle for a running interim-throughput monitor thread.
pub struct Monitor {
    stop:   Arc<AtomicBool>,
    handle: Option<JoinHandle<()>>,
}

impl Monitor {
    /// Stop the monitor and wait for the thread to finish.
    pub fn stop(mut self) {
        self.stop.store(true, Ordering::Relaxed);
        if let Some(h) = self.handle.take() { let _ = h.join(); }
    }
}

/// Reset the progress counter and, in gui mode, spawn a thread that emits a
/// `DOWNLOAD_RESULT`/`UPLOAD_RESULT` every 250 ms until stopped. A no-op monitor
/// is returned when gui mode is disabled.
pub fn spawn_transfer_monitor(dir: Direction, duration_secs: u32, open_test_uuid: Option<String>) -> Monitor {
    PROGRESS.store(0, Ordering::Relaxed);
    let stop = Arc::new(AtomicBool::new(false));

    let handle = if enabled() {
        let stop2 = stop.clone();
        Some(thread::spawn(move || {
            let start = Instant::now();
            while !stop2.load(Ordering::Relaxed) {
                thread::sleep(Duration::from_millis(250));
                let secs = start.elapsed().as_secs_f64();
                if secs <= 0.0 { continue; }
                let bytes = PROGRESS.load(Ordering::Relaxed);
                let mbps  = bytes as f64 * 8.0 / secs / 1_000_000.0; // decimal Mbit/s
                let frac  = (secs / duration_secs.max(1) as f64).min(1.0);
                transfer_result(dir, bytes, mbps, open_test_uuid.as_deref(), frac);
            }
        }))
    } else {
        None
    };

    Monitor { stop, handle }
}

File: clientJava/src/main/java/at/rtr/rmbt/client/Gui.java
package at.rtr.rmbt.client;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.concurrent.atomic.AtomicLong;

/**
 * Machine-readable JSON progress interface for the open-rmbt-desktop app.
 *
 * <p>See {@code doc/json_interface.md} for the specification. All JSON output is
 * gated by the {@code -v} flag; when disabled these methods are no-ops (except
 * the progress counter, a cheap atomic used by the interim monitor).
 *
 * <p>Throughput is reported in decimal Mbit/s ({@code bytes * 8 / s / 1e6}),
 * ping RTTs in ms, and {@code pingTimeNs} in ns.
 */
final class Gui {

    private static final ObjectMapper JSON = new ObjectMapper();
    private static volatile boolean enabled = false;
    /** Cumulative bytes transferred in the current transfer phase (all threads). */
    private static final AtomicLong PROGRESS = new AtomicLong(0);

    private Gui() {}

    static void setEnabled(boolean v) { enabled = v; }
    static boolean enabled() { return enabled; }

    private static void emit(ObjectNode n) {
        if (!enabled) return;
        try { System.out.println(JSON.writeValueAsString(n)); }
        catch (Exception ignored) {}
    }

    // ── Plain-text lifecycle sentinels (backward-compat shim, gui mode only) ────

    static void startingTest() { if (enabled) System.out.println("STARTING TEST."); }
    static void endingTest()   { if (enabled) System.out.println("ENDING TEST."); }

    // ── Messages ────────────────────────────────────────────────────────────

    static void stateChange(String state) {
        ObjectNode n = JSON.createObjectNode();
        n.put("type", "STATE_CHANGE");
        n.put("time", System.currentTimeMillis());
        n.put("state", state);
        emit(n);
    }

    static void uuidInfo(String testUuid, String openTestUuid, String token) {
        ObjectNode n = JSON.createObjectNode();
        n.put("type", "UUID_INFO");
        n.put("testUuid", testUuid);
        n.put("openTestUuid", openTestUuid);
        n.put("testToken", token);
        emit(n);
    }

    static void pingResult(long clientNs, long serverNs, long timeNs) {
        ObjectNode n = JSON.createObjectNode();
        n.put("type", "PING_RESULT");
        n.put("time", System.currentTimeMillis());
        n.put("phase", "PING");
        n.put("pingTimeNs", timeNs);
        n.put("pingClient", clientNs / 1e6); // ns → ms
        n.put("pingServer", serverNs / 1e6); // ns → ms
        n.put("status", "PING");
        emit(n);
    }

    private static void transferResult(boolean upload, long bytes, double mbps,
                                       String openTestUuid, double progress) {
        ObjectNode n = JSON.createObjectNode();
        n.put("type", upload ? "UPLOAD_RESULT" : "DOWNLOAD_RESULT");
        n.put("time", System.currentTimeMillis());
        n.put("phase", upload ? "UP" : "DOWN");
        n.put("down", upload ? -1.0 : mbps);
        n.put("up", upload ? mbps : -1.0);
        n.put("bytes", bytes);
        if (openTestUuid != null) n.put("testOpenUuid", openTestUuid);
        n.put("progress", progress);
        n.put("status", upload ? "UP" : "DOWN");
        emit(n);
    }

    // ── Interim throughput monitor ────────────────────────────────────────────

    static void addProgress(long bytes) { PROGRESS.addAndGet(bytes); }

    /**
     * Reset the progress counter and, in gui mode, start a daemon thread emitting
     * a DOWNLOAD_RESULT/UPLOAD_RESULT every 250 ms until {@link Monitor#stop()}.
     */
    static Monitor startTransferMonitor(boolean upload, int durationSecs, String openTestUuid) {
        PROGRESS.set(0);
        Monitor m = new Monitor(upload, durationSecs, openTestUuid);
        if (enabled) m.start();
        return m;
    }

    static final class Monitor {
        private final boolean upload;
        private final int     durationSecs;
        private final String  openTestUuid;
        private volatile boolean stop = false;
        private Thread thread;

        Monitor(boolean upload, int durationSecs, String openTestUuid) {
            this.upload = upload;
            this.durationSecs = durationSecs;
            this.openTestUuid = openTestUuid;
        }

        void start() {
            final long startNs = System.nanoTime();
            thread = new Thread(() -> {
                while (!stop) {
                    try { Thread.sleep(250); } catch (InterruptedException e) { break; }
                    double secs = (System.nanoTime() - startNs) / 1e9;
                    if (secs <= 0) continue;
                    long bytes  = PROGRESS.get();
                    double mbps = bytes * 8.0 / secs / 1e6; // decimal Mbit/s
                    double frac = Math.min(1.0, secs / Math.max(1, durationSecs));
                    transferResult(upload, bytes, mbps, openTestUuid, frac);
                }
            }, "gui-transfer-monitor");
            thread.setDaemon(true);
            thread.start();
        }

        void stop() {
            stop = true;
            if (thread != null) {
                thread.interrupt();
                try { thread.join(500); } catch (InterruptedException ignored) {}
            }
        }
    }
}
